_posts/2022-03-15-turn-based-offline-rl.md

```diff
@@ -0,0 +1,344 @@
+there, without performing any further steps.
+
+### Does the agent improve its performance over time?
+
+How many turns are needed to start having results comparable to
+the optimal policy? In other words, how much better are we performing if compared to not doing any turn at all?
+
+Let’s analyze the plot below. In this figure we are plotting the algorithm’s performance, measured in
+“averaged reward” (the higher, the better), as the amount of data available offline increases. In general, we expect
+the curve obtained by running the optimal policy (violet curve) to represent an upper bound: it’s the best we
+can achieve! On the other hand, we expect the curve obtained by running the random policy without any “turn”
+(green curve), to be our lower bound. Also, generally speaking, we expect that the more offline data is available,
+the better the achieved scores will be, since our chosen Offline RL will have more data coverage and possibility
+to converge to the optimal policy. Given this, we can observe that the performance of the turn-based procedure falls
+in the middle between the aforementioned upper and lower bounds: as the number of turns increases, the closer we get
+to the upper bound. However, we can observe that 3 turns are already enough to start having better performance than
+the lower bound. This plot confirms our hypothesis: “Turn-based Offline RL” stands exactly between Online RL
+(upper bound) and Offline RL (lower bound).
+
+![Figure 6](/img/articles/2022-03-15-turn-based-offline-rl/learning_curves_reduced.png)
+*Fig.6 — This plot shows the comparison between baselines and the turn-based procedure in terms of average reward (the higher,
+the better) as the size of the collected data used to train the algorithm offline grows. To obtain this figure,
+we have run each of the algorithms for 30 seeds. For the optimal policy we run Q-iteration, while for the rest we
+applied CQL with fixed alpha=0. For each run, CQL was fitted for 300 iterations.*
+
+## Conclusions and Future Work
+
+In this blog post we have presented a practical approach you could use to address cases where you have temporary
+and limited access to an environment, and you have computational resources at your disposal to train your RL algorithm
+“offline” only.
+
+In fact, the proposed solution falls in the middle between Online RL and Offline RL: our agent is warmed up
+by training it via Offline RL on a dataset generated by running the uniform random policy and then
+subsequently improved by accessing the environment in “turns”, thus partially simulating what you would get on a
+standard online RL scenario.
+
+In particular, we show that:
+
+* The turn-based procedure is effective since the policy learned in subsequent turns consistently improves as
+turns increase, matching the expectations. This result is demonstrated through some visualizations, showing how
+the agent chooses a better and faster path to target turn after turn.
+
+
```

Comment:
zbędne puste linie rozbiją listę bullet pointów na dwie osobne

Reply:
All right, I've removed them here and also in other places of the file.